keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -91,242 +92,324 @@ with the unreasonable development overhead of a webhook.
 
 ```yaml
 properties:
   foo:
-    type: array
     items:
-      x-kubernetes-immutable: true
-      type: string
+      type: object
+      properties: ...
+      x-kubernetes-mutability: Immutable
 ```
 
-means that the `foo` array itself is mutable, but the items are not. This means that
-items can be deleted and added, but at each existing index the string values are
-immutable.
+Allowed:
+- `{}` → `{foo:...}`
+- `{foo:...}` → `{}`
+- `{foo:[a]}` → `{foo:[a,b]}` (value of key 0 is unchanged)
+- `{foo:[a]}` →	`{foo:[]}`
+
+Disallowed:
+- `{foo:[a]}` → `{foo:[b]}`
+- `{foo:[a]}` → `{foo:[b,a]}` (value of key 0 is changed)
+
+#### Mutable array, immutable items, list type map
+
+The key-value pairs in the array are immutable, set set of keys is not. Hence, addition and removal of key-values pairs
+
+*Example 3*:
 
 ```yaml
 properties:
   foo:
-    type: object
-    x-kubernetes-immutable: true
+    x-kubernetes-list-type: map
+    x-kubernetes-list-map-keys: ["k"]
+    items:
+      type: object
+      properties: ...
+      x-kubernetes-mutability: Immutable
 ```
 
-means that the whole `.foo` object is immutable. It cannot be removed or set.
+Allowed:
+- `{}` → `{foo:...}`
+- `{foo:...}` → `{}`
+- `{foo:[{k:a}]}` → `{foo:[{k:a},{k:b}]}` (value of key a is unchanged)
+- `{foo:[{k:a}]}` → `{foo:[]}`
+- `{foo:[{k:a}]}` → `{foo:[{k:b}]}`
+- `{foo:[{k:a},{k:b}]}`	→ `{foo:[{k:b},{k:a}]}`	(values of key a and b unchanged)
+
+Disallowed:
+- `{foo:[{k:a,v:1}]}` → `{foo:[{k:a,v:2}]}` (value of key a is changed)
+
+**TODO:** removal+addition = change. The former is allowed, the latter disallowed. 
+
+#### Mutable array, immutable items, list type set
 
-If one wants to make only field immutable, but allow to delete an existing object
-or set it if unset, the following is used:
+Sets are maps with the whole value as key. Hence, addition and removed of any value in the set is allowed.
+
+*Example 4*:
 
 ```yaml
 properties:
   foo:
-    type: object
-    properties:
-      "x":
-        type: string
-        x-kubernetes-immutable: true
-      "y":
-        type: string
-        x-kubernetes-immutable: true
+    x-kubernetes-list-type: set
+    items:
+      type: string
+      x-kubernetes-mutability: Immutable
 ```
 
-for all fields `x` and `y` of `foo`.
+Allowed:
+- `{}` → `{foo:...}`
+- `{foo:...}` → `{}`
+- `{foo:[a]}` → `{foo:[a,b]}` (value of key a is unchanged)
+- `{foo:[a]}` → `{foo:[]}` (key a is removed)
+- `{foo:[a]}` → `{foo:[b]}` (key a is removed, key b is added)
+- `{foo:[a]}` → `{foo:[b,a]}` (value of key a is unchanged, key b is added)
+
+#### Mutable map, immutable values, map type map
+
+Equivalently to the list type map, removal and addition of key-value pairs are allowed, while direct value change is not.
+
+*Example 5*:
 
 ```yaml
 properties:
   foo:
-    type: object
-    x-kubernetes-immutable: true
-    properties:
-      bar:
-        type: object
-        x-kubernetes-immutable: true
+    x-kubernetes-list-type: map
```

Comment:
why  `x-kubernetes-list-type: map` here?
Shouldn't it be `x-kubernetes-map-type: granular`? or just `type: object`?

Reply:
yes